Introduce PSR-7 `ResponseInterface` for Modules executions

nupplaphil opened the pull request:
FollowUp #11002 

Split and delete `ModuleController`:
- $moduleName is part of the argument string => App\Arguments
- $isBackend boolean already part of App\Mode::isBackend()
- $module is now the direct return of App\Router::getModule()
- ModuleController::run() moved to BaseModule::run()

Introduce `Response` class and corresponding interfaces to create a standard PSR-7 compliant `ResponseInterface`
This particular interface (`ResponseInterface`) withis now returned after each `Module` call (including the `API` calls) and makes all `Module`s testable

Needs https://github.com/friendica/friendica-addons/pull/1209 as well for rendertime :-)

nupplaphil committed:
Split and delete `ModuleController`
- $moduleName is part of the argument string => App\Arguments
- $isBackend boolean already part of App\Mode::isBackend()
- $module is now the direct return of App\Router::getModule()
- ModuleController::run() moved to BaseModule::run()

MrPetovan commented on src/App.php:
```diff
@@ -703,20 +700,20 @@ public function runFrontend(App\ModuleController $module, App\Router $router, IM
 
 			if (!$this->mode->isInstall() && !$this->mode->has(App\Mode::MAINTENANCEDISABLED)) {
-				$module = new ModuleController('maintenance', new Maintenance());
+				$module = new Maintenance();
```
Needs `L10n` instance as per your recent fixings.

MrPetovan commented on src/App.php:
```diff
@@ -703,20 +700,20 @@ public function runFrontend(App\ModuleController $module, App\Router $router, IM
 				// determine the module class and save it to the module instance
 				// @todo there's an implicit dependency due SESSION::start(), so it has to be called here (yet)
-				$module = $module->determineClass($this->args, $router, $this->config, $dice);
+				$module = $router->getModule();
```
What about legacy modules still in `mod/`?

nupplaphil replied:
It's now moved directly into the router. The router now didn't "just" find the right parameters & classes, it also creates the class it with the help of `Dice`, see the new `Router::getModule()` method

MrPetovan commented on src/App/Mode.php:
```diff
@@ -25,6 +25,7 @@
 use Friendica\Core\Config\ValueObject\Cache;
 use Friendica\Database\Database;
 use Friendica\Util\BasePath;
+use phpDocumentor\Reflection\Types\Static_;
```
I'm sorry what?

nupplaphil replied:
hehe, sometimes I start typing `stat..` and hit the "Tab" Button, and voliá, the IDE thinks importing `Static_` seems the right guess ;-)

nupplaphil replied:
normally I don't forget to remove it again .. not this time :D

MrPetovan commented on src/App/Router.php:
```diff
@@ -83,6 +91,15 @@ class Router
 	/** @var ICanLock */
 	private $lock;
 
+	/**	@var Arguments */
```
```suggestion
	/** @var Arguments */
```

MrPetovan commented on src/App/Router.php:
```diff
@@ -91,14 +108,21 @@ class Router
 	 * @param string              $baseRoutesFilepath The path to a base routes file to leverage cache, can be empty
 	 * @param L10n                $l10n
 	 * @param ICanCache           $cache
+	 * @param ICanLock            $lock
```
This parameter doesn't exist.

nupplaphil replied:
of course it does o_O see https://github.com/friendica/friendica/pull/11005/files/a7a05633d8226edb76eb9a9f0bce184db82bd4a3#diff-7f7aabf2d5c355cac62f512a0d5387e00dfcb00b27e6dcb39b3b90ff028d05dfR122

MrPetovan reviewed (changes requested):
I like all the deletions but so far this PR doesn't pass the checks.

nupplaphil committed:
Add feedback :)

nupplaphil committed:
Make `BaseModule` a real entity
- Add all dependencies, necessary to run the content (baseUrl, Arguments)
- Encapsulate all POST/GET/DELETE/PATCH/PUT methods as protected methods inside the BaseModule
- Return Module content ONLY per `BaseModule::run()` (including the Hook logic there as well)

nupplaphil commented:
Another big PR, but we're on a good way to test any module we want without mocking/spying/..

Tests are not in my focus yet, I'll check them later :D

-> it's now running at https://friendica.philipp.info
-> I'll check for any errors (post/preview/delete still working ;-) )

nupplaphil committed:
Fix LegacyModule content return

nupplaphil committed:
Introduce `Response` for Modules to create a testable way for module responses

nupplaphil committed:
Inherit `ApiResponse` from `Response`

nupplaphil committed:
Add NodeInfo Module test as an example

nupplaphil commented:
Ok, I think I leave it for now...
... Testing the whole branch for at least another day 

At least workers are running fine so far and I can post/update/delete , see every profile, can edit permissions, ...

But **why** did I change that much:
See https://github.com/friendica/friendica/pull/11005/commits/9494e2e4b389c8c2f69304bb3782c71fdbd99537

We can now do tests for complete modules :-)
With the help of `Response` every module now returns an interface with:
- `getContent()` - A mixed content, set by the Module (the core output)
- `getType()` - The type of content (HTML, RSS, ...)
- `getHeaders()` - An array with headers, we **would** set

Currently, almost every Module directly uses `header()` and `System::*()` calls to exit directly, but we can now replace this calls step by step directly with new method calls `$this->response->setHeader(string $header, ?string $key)` and `$this->response->addContent($content)` and add a test for every transformed module if we want to ;-)

nupplaphil committed:
Replace `header()` with `$response->setHeader()` at `BaseModule`

nupplaphil committed:
Delete deprecated ApiResponseDouble

MrPetovan commented:
While we're at it, can we use [PSR-7 compatible messages](https://www.php-fig.org/psr/psr-7/)? We already depend on the Guzzle implementation, so we can directly use `GuzzleHttp\Psr7\Response` without needing an additional dependency.

nupplaphil committed:
Replace IRespondToRequests with PSR-7 ResponseInterface

nupplaphil committed:
Fixing Response

nupplaphil committed:
Fixing Response

nupplaphil committed:
Fix Page Header usage

nupplaphil committed:
Update messages.po

nupplaphil commented:
No errors so far in my logs for a week ..

--> I rebased the changes of this week into this branch and start a smoke test again :)

nupplaphil commented:
The next PR will cover the `HTTPInputData` class to make the whole API changes from @annando testable again (I tried it to add it at this PR, but as said I need to change the HTTPInput behavior) :-)

nupplaphil committed:
Fix Contact modules

nupplaphil committed:
Fix that (raw)content is always executed during Module::run()

nupplaphil committed:
Fixing https://github.com/friendica/friendica/issues/11033

MrPetovan commented on src/Module/Contact/Profile.php:
```diff
@@ -355,7 +355,7 @@ protected function content(array $request = []): string
 			'$pending'                   => ($contact['pending'] ? $this->t('Awaiting connection acknowledge') : ''),
 			'$hidden'                    => ['hidden', $this->t('Hide this contact from others'), $localRelationship->hidden, $this->t('Replies/likes to your public posts <strong>may</strong> still be visible')],
-			'$notify'                    => ['notify', $this->t('Notification for new posts'), ($contact['notify_new_posts'] == 1), $this->t('Send a notification of every new post of this contact')],
+			'notify_new_posts'           => ['notify_new_posts', $this->t('Notification for new posts'), ($localRelationship->notifyNewPosts), $this->t('Send a notification of every new post of this contact')],
```
Please respect the `$` notation for template variables, both work but every other key in the code is prefixed with a dollar sign.

nupplaphil committed:
Respect the Dollar(notation :-))

nupplaphil committed:
Fix Tos Module

nupplaphil commented:
so, now I hopefully fixed everything.
Note to myself: Dice doesn't pass arguments from `$dice->create(Class::class, $arguments)` to any other class except the initial class. Means that `Tos` didn't got the original parameters when adding as implicit dependency at the constructor :-/
